### ISSUE TITLE
Broken TXT lookup with etcd backend

### ISSUE THREAD

**Issue by heilerich** [miekg/coredns#5283]
**What happened**:

Setting a TXT records with the etcd plugin, e.g. as per documentation with

```
etcdctl put /skydns/local/skydns/x6 '{"ttl":60,"text":"this is a random text message."}'
```

returns NXDOMAIN.

**What you expected to happen**:

The TXT record is returned.

**How to reproduce it (as minimally and precisely as possible)**:

Follow the documentation to setup the etcd plugin and add the TXT record as per documentation.

**Anything else we need to know?**:

This bug was introduced in https://github.com/coredns/coredns/commit/e3266d24f7a3a120739875ff421d2d47d044ad69 when the `exact` argument to the Service function was changed from false to true [here](https://github.com/coredns/coredns/commit/e3266d24f7a3a120739875ff421d2d47d044ad69#diff-9e2e0397d9ac3d088c20f0d3a8058af9aa4a5eb3a5c6373e3e7fafe10ee7fdcaR330). This has two effects:

1. Only TXT records that exactly match the domain are returned (hence the non-functional example in the documentation with uses the `x6` subkey)
2. It is not possible anymore to set multiple TXT records on one domain

This bug was actually caught by the test suite and reported in #3810 but instead of fixing the issue, the test was "fixed".

**Environment**:

- the version of CoreDNS:
- Corefile:

```
. {
    etcd {
        path /skydns
        endpoint http://etcd:2379
    }
    log
}
```

**Comment by heilerich** [comment]
I think the underlying issue here is the behavior of the etcd plugin when asked to lookup records only providing **exact** matches.

Let's say we are trying to lookup an exact match for `test.skydns.local`

Currently, the plugin only considers one key `/prefix/local/skydns/test`, i.e. it is doing an ETCD `get` without the prefix flag. But the record at `/prefix/local/skydns/test/` should also be considered an exact match.

Additionally, the only supported way of providing multiple records using skyDNS semantics is using subkeys, that is in this case `/prefix/local/skydns/test/txt1` `.../txt2` etc. But these keys are ambiguous: Are they supposed to be multiple entries for `test.skydns.local` or are they actually separate records at `txt1.test.skydns.local` etc.?

As far as I can tell the only way of providing an 'exact' match with multiple records with skyDNS semantics is using the `group` attribute. In the example case this would mean creating the following entries

```
/prefix/local/skydns/test/   '{"ttl":60,"text":"text message one","group":"test.skydns.local"}'
/prefix/local/skydns/test/t1 '{"ttl":60,"text":"text message two","group":"test.skydns.local"}'
```

So I would suggest the following changes:

(1) When the `Services` (from the `ServiceBackend` interface) function is called with `exact=true` it should get all the matching Records (without disabling recursion).

(2) After grouping the results if exact is true, check if the first record's name (which should be the shortest key) matches the question's name. If it matches and the record has a group return all entries in the group (discarding entries with an empty group attribute). If it matches and the record has no group, return only that record. If the first name does not match, return an empty slice.

Additionally, the `TXT` function in the `plugin` package must be updated, because it currently stops iteration on the first returned TXT record.

This shouldn't have any impact on the current behavior other than allowing multiple TXT records using the grouping semantic.

Lastly, the documentation must be updated to reflect this and the above mentioned test case for TXT records stored in subkeys should be re-added.

I would be willing to implement these changes, but I would greatly appreciate any feedback regarding the proposed strategy from a maintainer before doing so.

**Comment by chrisohaver** [comment]
> the only supported way of providing multiple records using skyDNS semantics is using subkeys, that is in this case /prefix/local/skydns/test/txt1 .../txt2 etc. But these keys are ambiguous: Are they supposed to be multiple entries for test.skydns.local or are they actually separate records at txt1.test.skydns.local etc.?

They represent both separate records and multiple at the parent domains at the same time.  AFAIK, this is how SkyDNS designed to work (it's called "recursion" in SkyDNS docs - not to be confused with DNS recursion).

See #5293 for my proposed fix.

**Comment by heilerich** [comment]
Great! Many thanks @chrisohaver for the effort. This seems to solve the problem of setting multiple TXT records.

On a side note: While looking at your PR, I just realized that it is possible to get CoreDNS to reply with multiple CNAME for the same name (because the iteration is not stopped on the first encountered CNAME). This should not be possible, because obviously there cannot be multiple canonical names for a name. Additionally, it is also possible now to have a TXT record *and* a CNAME for the same name, which technically should also not be possible. [Reference (RFC 1034 Section 3.6.2)](https://datatracker.ietf.org/doc/html/rfc1034#section-3.6.2)

Other records types are also affected. E.g. this also happens with A records on the current release.

Set the following records
```
/skydns/test/zone/sub/c1 '{"host":"canon.zone.test","ttl":60,"group":"a"}'
/skydns/test/zone/sub/c2 '{"host":"alternative-reality.zone.test","ttl":60,"group":"a"}'

/skydns/test/zone/canon '{"host":"1.1.1.1","ttl":60}'
/skydns/test/zone/alternative-reality '{"host":"1.1.1.2","ttl":60}'

/skydns/test/zone/sub/c3 '{"host":"1.1.1.3","ttl":60,"group":"a"}'
```
and CoreDNS will reply with

```
;; ANSWER SECTION:
sub.zone.test.				60	IN	CNAME	canon.zone.test.
canon.zone.test.			60	IN	A	1.1.1.1
sub.zone.test.				60	IN	CNAME	alternative-reality.zone.test.
alternative-reality.zone.test.		60	IN	A	1.1.1.2
sub.zone.test.				60	IN	A	1.1.1.3
```

I think this could be fixed by stopping iteration on the first encountered CNAME. And maybe the existing records that have been added to the `records` slice must also be cleared when a CNAME is encountered. Although, I am not sure how to handle this ... What should have precedence here: The CNAME or any TXT/A/MX records on the same name ... Probably whatever has the first shortest key.

Since this affects other record types as well maybe this warrants a separate issue?

**Comment by chrisohaver** [comment]
> Since this affects other record types as well maybe this warrants a separate issue?

It is a different issue than this one, so it would be a separate issue.
Although, I think for that issue we can just take a "garbage in garbage out" position.  Maybe document it as a known caveat.

**Comment by heilerich** [comment]
Sounds reasonable. Seeing that the current documentation is missing a section on CNAMEs in general, adding a short explanation would be helpful I guess.

On the other hand, I am always in favor of checking for known common errors :) At least emitting a warning somewhere if illegal records are detected. But since this is illegal behavior in general the etcd plugin is probably not the right place to do this anyways. Other DNS servers, I have experience with, will refuse to load zones that contain these errors …
That being said this is not a problem for me at this point, so I am perfectly happy once your PR is merged.

Thanks again for the fast fix!
